### PR TITLE
Implement Azure Identity handling for auth

### PR DESCRIPTION
**Pull request by tostringtheory**
I updated the Azure Sign Tool to support an Azure Managed Identity.  This allows utilizing the local developer credentials on a machine running the tool.

### CODE REVIEW

File: src/AzureSign.Core/AzureSign.Core.csproj [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
-    <Description>Authenticode signing library.</Description>
-    <GenerateDocumentationFile>true</GenerateDocumentationFile>
-    <RuntimeIdentifiers>win10-x64;win-10-x86</RuntimeIdentifiers>
-    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
-  </PropertyGroup>
+	<PropertyGroup>
+		<TargetFramework>netcoreapp2.1</TargetFramework>
+		<Description>Authenticode signing library.</Description>
+		<GenerateDocumentationFile>true</GenerateDocumentationFile>
+		<RuntimeIdentifiers>win10-x64;win10-x86</RuntimeIdentifiers>
```
Comment (vcsjones) [inline src/AzureSign.Core/AzureSign.Core.csproj]: The RID fix is good, thank you.

File: src/AzureSignTool/SignCommand.cs [hunk 2]
```diff
@@ -345,6 +350,9 @@ private static ErrorOr<X509Certificate2Collection> GetAdditionalCertificates(IEn
             return collection;
         }
 
-        private static bool OneTrue(bool left, bool right) => left != right && (left || right);
+        private static bool OneTrue(params bool[] checks)
+        {
+	        return checks.Count(boolean => boolean == true) == 1;
```
Comment (vcsjones) [inline src/AzureSignTool/SignCommand.cs]: This could probably be simplified to:
```suggestion
	        return checks.Count(b => b) == 1;
```

Comment (tostringtheory) [reply src/AzureSignTool/SignCommand.cs]: I had actually written the line exactly like you had it to begin with.  I was concerned with readability/clarity for future coders, so expanded it out to be clear that it's counting the true booleans.  But yeah, I can re-condense the change.

**Comment by tostringtheory** [comment]
@vcsjones - Man, sorry the whitespace skipped through, thought I had that.  I'll clean that up and implement your one change and send a PR later today/this week.

**Comment by ghstahl** [comment]
Since we now have 3 ways to authenticate, can we think about the flag being an enum.
i.e. --authentication-type = <token|client_credentials|managed_identity>
this way it is very explicit as to the one that will be used.

**Comment by tostringtheory** [comment]
@vcsjones - apologies for the delay on this, the you-know-what ended up throwing things way out of whack with work, and this slipped my mind while I was putting out fires there.  I was getting back to looking at this, and while I would agree that a flag/enum (re @ghstahl) is definitely the cleanest way to do things, the reason that I choose NOT to do that originally is basically to maintain backwards compatibility with any CI/CD or anyones knowledge of the tool as of now.  

As well, I have no experience with the command line library that you are using and not much free time to implement that as a full enum option now.  Before I go ahead and make the original changes requested above and get that resubmitted, I wanted to check and see if you still want them in this state, or if you'd prefer for me to abandon this PR and you/someone else take a further crack at implementing a flags option.

**Comment by clairernovotny** [comment]
We should be using the newer `Azure.Identity` way of doing MSI/App Auth instead of using the AppAuth library.

**Comment by tostringtheory** [comment]
> We should be using the newer `Azure.Identity` way of doing MSI/App Auth instead of using the AppAuth library.

Agreed - if @vcsjones wants to continue ahead with this PR (with changes), I'll switch over to the newer library.  Thanks for reminding me on that too - I've switched over most of my company's internal stuff to Azure.Identity already, not sure if there was a reason I didn't use that here at the time, but I can check on that.

**Comment by ghstahl** [comment]
@tostringtheory I am ok with your original implementation as having MSI is a win

**Comment by clairernovotny** [comment]
I got Managed Identity and the switchover to Azure Identity working in `RSAKeyVaultProvider`.

You can see the relevant files (KeyVaultConfiguraitonDiscoverer, AzureKeyVaultSignConfigurationSet) here:
https://github.com/novotnyllc/RSAKeyVaultProvider/tree/master/RSAKeyVaultProvider.Tests

Couple things to note: needs to be Azure.Identity 1.2.0-preview.2 or later for MSI to work in VS. You'll also need `RSAKeyVaultProvider` 2.0.15 or later. There's one more breaking change -- `TenantId` is now required to use a service principal.

**Comment by tostringtheory** [comment]
I finally was able to get some time this weekend to get back to this.  I rebased onto current `master` and applied your fixes above.  I held off on a full transition to the new `Azure.Identity` package since the solution is still utilizing the legacy `RSAKeyVaultProvider`.  My time is still limited and I wanted to get this in place first, before going down the rabbit hole of redoing auth.

**Comment by clairernovotny** [comment]
I have this all done here https://github.com/vcsjones/AzureSignTool/pull/83

**Comment by tostringtheory** [comment]
> I have this all done here #83

Oh wow, fantastic!  I think that supersedes this PR then?  If so I can abandon it.

**Comment by clairernovotny** [comment]
Yup, just need @vcsjones to merge it ;)

**Pull request closed without merging**